src/main/scala/org/apache/spark/sql/redis/RedisSourceRelation.scala

```diff
@@ -237,42 +239,34 @@ class RedisSourceRelation(override val sqlContext: SQLContext,
   /**
     * read rows from redis
     */
-  private def scanRows(node: RedisNode, keys: Seq[String], requiredColumns: Seq[String]): Seq[Row] = {
-    def filteredSchema(): StructType = {
-      val requiredColumnsSet = Set(requiredColumns: _*)
-      val filteredFields = schema.fields
-        .filter { f =>
-          requiredColumnsSet.contains(f.name)
-        }
-      StructType(filteredFields)
-    }
-
+  private def scanRows(node: RedisNode, keys: Seq[String],
+                       requiredColumns: Seq[String]): Seq[Row] = {
     val conn = node.connect()
 
     val pipelineValues = mapWithPipeline(conn, keys) { (pipeline, key) =>
       persistence.load(pipeline, key, requiredColumns)
     }
-
-    val rows =
-      if (requiredColumns.isEmpty || persistenceModel == SqlOptionModelBinary) {
-        pipelineValues
-          .map {
-            case jmap: JMap[_, _] => jmap.toMap
-            case value: Any => value
-          }
-          .map { value =>
-            persistence.decodeRow(value, schema, inferSchemaEnabled)
-          }
-      } else {
-        pipelineValues.map { case values: JList[_] =>
-          val value = requiredColumns.zip(values.asInstanceOf[JList[String]]).toMap
-          persistence.decodeRow(value, filteredSchema(), inferSchemaEnabled)
-        }
-      }
+    val rows = keys.zip(pipelineValues).map { case (key, value) =>
+      val keyMap = keyName -> tableKey(keysPrefixPattern, key)
+      persistence.decodeRow(keyMap, value, filteredSchema(requiredColumns),
```

Comment:
We can optimize it executing filteredSchema() only once for all keys.

Reply:
It's already cached in general case. It's little tricky with the `inferSchema` option. However, I think it's still possible to optimize more.